apps/serdes/Deserializer.cpp

```diff
@@ -0,0 +1,608 @@
+        auto predicate = deserialize_expr(load_expr->predicate_type(), load_expr->predicate());
+        auto index = deserialize_expr(load_expr->index_type(), load_expr->index());
+        return Halide::Internal::Load::make(Halide::Int(64), name, index, Halide::Buffer<float, 3>(), Halide::Internal::Parameter(), predicate, Halide::Internal::ModulusRemainder());
+    }
+    case Halide::Serdes::Expr::Expr_Ramp: {
+        const Halide::Serdes::Ramp *ramp_expr = (const Halide::Serdes::Ramp *)expr;
+        auto base = deserialize_expr(ramp_expr->base_type(), ramp_expr->base());
+        auto stride = deserialize_expr(ramp_expr->stride_type(), ramp_expr->stride());
+        auto lanes = ramp_expr->lanes();
+        return Halide::Internal::Ramp::make(base, stride, lanes);
+    }
+    case Halide::Serdes::Expr::Expr_Broadcast: {
+        const Halide::Serdes::Broadcast *broadcast_expr = (const Halide::Serdes::Broadcast *)expr;
+        auto value = deserialize_expr(broadcast_expr->value_type(), broadcast_expr->value());
+        auto lanes = broadcast_expr->lanes();
+        return Halide::Internal::Broadcast::make(value, lanes);
+    }
+    case Halide::Serdes::Expr::Expr_Let: {
+        const Halide::Serdes::Let *let_expr = (const Halide::Serdes::Let *)expr;
+        auto name = deserialize_string(let_expr->name());
+        auto value = deserialize_expr(let_expr->value_type(), let_expr->value());
+        auto body = deserialize_expr(let_expr->body_type(), let_expr->body());
+        return Halide::Internal::Let::make(name, value, body);
+    }
+    case Halide::Serdes::Expr::Expr_Call: {
+        const Halide::Serdes::Call *call_expr = (const Halide::Serdes::Call *)expr;
+        auto name = deserialize_string(call_expr->name());
+        auto args_type = call_expr->args_type();
+        auto args_serialized = call_expr->args();
+        std::vector<Halide::Expr> args;
+        for (size_t i = 0; i < args_serialized->size(); ++i) {
+            auto arg = deserialize_expr(args_type->Get(i), args_serialized->Get(i));
+            args.push_back(arg);
+        }
+        auto value_index = call_expr->value_index();
+        // TODO: fix type and function ptr here once function DAG is fixed
+        return Halide::Internal::Call::make(Halide::Int(64), name, args, Halide::Internal::Call::CallType::Extern, Halide::Internal::FunctionPtr(), value_index);
+    }
+    case Halide::Serdes::Expr::Expr_Variable: {
+        const Halide::Serdes::Variable *variable_expr = (const Halide::Serdes::Variable *)expr;
+        auto name = deserialize_string(variable_expr->name());
+        return Halide::Internal::Variable::make(Halide::Int(64), name);
+    }
+    case Halide::Serdes::Expr::Expr_Shuffle: {
+        const Halide::Serdes::Shuffle *shuffle_expr = (const Halide::Serdes::Shuffle *)expr;
+        auto vectors_type = shuffle_expr->vectors_type();
+        auto vectors_serialized = shuffle_expr->vectors();
+        std::vector<Halide::Expr> vectors;
```

Comment:
Here and elsewhere: seems like writing a `deserialize_expr_vector()` method would pay for itself.